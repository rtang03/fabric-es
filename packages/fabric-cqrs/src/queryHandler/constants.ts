export const INVALID_ARG = 'invalid input argument';
