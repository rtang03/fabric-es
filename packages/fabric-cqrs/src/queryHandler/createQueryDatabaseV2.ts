import flatten from 'lodash/flatten';
import { Redisearch } from 'redis-modules-sdk';
import type { Commit, QueryDatabase } from '../types';
import { getLogger } from '../utils';

/**
 * @about Create query database
 * @params redis
 * @returns [[QueryDatabase]]
 */
export const createQueryDatabaseV2: (client: Redisearch) => QueryDatabase = (client) => {
  const logger = getLogger({ name: '[query-handler] createQueryDatabase.js', target: 'console' });
  const countNonNull = (deletedItems: number[][]) =>
    flatten(deletedItems)
      .filter((item) => !!item)
      .reduce((prev, curr) => prev + curr, 0);
  const getHistory = (commits: Commit[]): any[] => {
    const history = [];
    commits.forEach(({ events }) => events.forEach((item) => history.push(item)));
    return history;
  };

  return {
    clearNotification: null,
    deleteCommitByEntityId: null,
    deleteCommitByEntityName: null,
    getNotification: null,
    fullTextSearchCommit: null,
    fullTextSearchEntity: null,
    mergeCommit: null,
    mergeCommitBatch: null,
    mergeEntity: null,
    mergeEntityBatch: null,
    queryCommitByEntityId: null,
    queryCommitByEntityName: null,
    queryEntity: null,
  };
};
