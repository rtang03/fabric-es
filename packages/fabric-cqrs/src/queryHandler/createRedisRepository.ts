import util from 'util';
import startsWith from 'lodash/startsWith';
import { FTCreateParameters, FTSchemaField, Redisearch } from 'redis-modules-sdk';
import type { Selector } from 'reselect';
import type { Commit } from '../types';
import { getLogger } from '../utils';
import { postSelector as restoreCommit } from './model';
import { pipelineExec } from './pipelineExec';
import type { FieldOption, RedisearchDefinition, RedisRepository } from './types';
import type { CommitInRedis } from './types';

/**
 * @about create abstract layer for redis repository
 */
export const createRedisRepository: <TInput, TItemInRedis, TOutput>(option: {
  client: Redisearch;
  kind?: 'entity' | 'commit';
  fields: RedisearchDefinition<TInput>;
  entityName: string;
  param?: FTCreateParameters;
  preSelector?: Selector<[TInput, Commit[]?], TItemInRedis>;
  postSelector?: Selector<TItemInRedis, TOutput>;
}) => RedisRepository<TOutput> = <TItem, TItemInRedis, TResult>({
  client,
  kind = 'entity' as any,
  fields,
  entityName,
  param,
  preSelector,
  postSelector,
}) => {
  const logger = getLogger({ name: '[query-handler] createRedisRepository.js', target: 'console' });

  // every entity is indexed with Prefix "e:entityName:". commit is "c:"
  const indexName = { entity: `eidx:${entityName}`, commit: 'cidx' }[kind];

  const prefix = { entity: `e:${entityName}:`, commit: 'c:' }[kind];

  // compute key
  const getKey = {
    entity: ({ id }) => (!id ? null : `${prefix}${id}`),
    commit: ({ entityName, entityId, commitId }: Commit) =>
      !entityName || !entityId || !commitId
        ? null
        : `${prefix}${entityName}:${entityId}:${commitId}`,
  }[kind];

  // add default FTCreateParameters
  const getParam: (param: FTCreateParameters) => FTCreateParameters = (param) =>
    Object.assign({}, { prefix: [{ count: 1, name: prefix }] }, param);

  // compute schema
  const getSchema: <F>(fields: F) => FTSchemaField[] = <F extends FieldOption>(input) =>
    Object.entries<F>(input)
      .map(
        ([key, { altName, index }]) =>
          // if index is required
          index && {
            ...index,
            // append the indexKey name
            name: altName ?? key,
          }
      )
      .filter((item) => !!item);

  return {
    createIndex: () => client.create(indexName, getSchema<TItem>(fields), getParam(param)),
    deleteItemsByPattern: async (pattern) => {
      try {
        return await pipelineExec<number>(client, 'DELETE', pattern).then((data) => [
          data.map(([err, _]) => err),
          data.map(([_, count]) => count).reduce((pre, cur) => pre + cur, 0),
        ]);
      } catch (e) {
        logger.error(util.format('fail to deleteCommitsByPattern, %j', e));
        return [e, null];
      }
    },
    dropIndex: (deleteHash = true) => client.dropindex(indexName, deleteHash),
    hmset: (item, history) => {
      const key = getKey(item);

      if (!key) throw new Error('invalid key');

      return client.redis.hmset(key, preSelector?.([item, history]) || item);
    },
    hgetall: (key) =>
      client.redis.hgetall(key).then((result) => (postSelector?.(result) || result) as TResult),
    getKey: (item: TItemInRedis) => getKey(item),
    getIndexName: () => indexName,
    getPattern: (pattern, args) =>
      ({
        COMMITS_BY_ENTITYNAME: `c:${args[0]}:*`,
        COMMITS_BY_ENTITYNAME_ENTITYID: `c:${args[0]}:${args[1]}:*`,
        ENTITIES_BY_ENTITYNAME: `e:${args[0]}:*`,
        ENTITIES_BY_ENTITYNAME_ENTITYID: `e:${args[0]}:${args[1]}:*`,
      }[pattern]),
    getPreSelector: <TInput, TOutput>(): Selector<TInput, TOutput> => preSelector,
    getPostSelector: <TInput, TOutput>(): Selector<TInput, TOutput> => postSelector,
    queryCommitsByPattern: async (pattern) => {
      try {
        return await pipelineExec<CommitInRedis>(client, 'HGETALL', pattern).then((data) => [
          data.map(([err, _]) => err),
          data.map(([_, commit]) => restoreCommit(commit)),
        ]);
      } catch (e) {
        logger.error(util.format('fail to queryCommitsByPattern, %j', e));
        return [e, null];
      }
    },
    search: async ({ countTotalOnly, kind, index, query, param, restoreFn }) => {
      try {
        const customParm = countTotalOnly
          ? { ...param, ...{ limit: { first: 0, num: 0 } } }
          : param;
        // step 1: use FT.SEARCH to find corresponding keys
        const data: any[] = await client.search(index, query, customParm);

        // Notice that if there is paginated result, the number of total count of search result, data[0] will the total count.
        // But the subsequment array element will only return ONE page
        // e.g.
        // [
        //   2, <== Number of total count of search result
        //   'c:test_proj:qh_proj_test_001:20200528133520841',
        //   [
        //     'entityName',
        //     'test_proj',
        //     'id',
        //    ....
        // ]
        const prefix = { commit: 'c:', entity: 'e:' }[kind];

        // keys will be ['c:test_proj:qh_proj_test_001:20200528133520841', /* ... */ ]
        const keys = data.slice(1).filter((item) => startsWith(item, prefix));

        const count = countTotalOnly ? data[0] : keys?.length;

        if (countTotalOnly) return [[], count, null];

        // step 2: retrieve actual content, and restore to proper shape
        return await pipelineExec<TResult>(client, 'HGETALL', null, keys).then((data) => [
          data.map(([err, _]) => err),
          count,
          data.map(
            ([_, item]) => ({ commit: restoreCommit, entity: restoreFn }[kind]?.(item) || item)
          ),
        ]);
      } catch (e) {
        logger.error(util.format('fail to search, %j', e));
        return [[e], null, null];
      }
    },
  };
};
