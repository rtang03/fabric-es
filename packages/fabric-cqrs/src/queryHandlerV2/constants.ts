export const INVALID_ARG = 'invalid input argument';
export const REDUCE_ERR = 'fail to reduce to currentState';
