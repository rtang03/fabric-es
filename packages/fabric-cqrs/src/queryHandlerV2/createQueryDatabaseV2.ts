import util from 'util';
import filter from 'lodash/filter';
import groupBy from 'lodash/groupBy';
import isEqual from 'lodash/isEqual';
import { FTSearchParameters, Redisearch } from 'redis-modules-sdk';
import { Commit, HandlerResponse, trackingReducer } from '../types';
import { getLogger, isCommit } from '../utils';
import {
  INVALID_ARG,
  NO_RECORDS,
  QUERY_ERR,
  REDIS_ERR,
  REDUCE_ERR,
  REPO_NOT_FOUND,
} from './constants';
import { commitSearchDefinition, postSelector, preSelector } from './model';
import type { CommitInRedis, OutputCommit, QueryDatabaseV2, RedisRepository } from './types';
import { createNotificationCenter, createRedisRepository } from '.';

/**
 * @about create query database
 * @params redisearch client
 * @returns [[QueryDatabase]]
 */
export const createQueryDatabaseV2: (
  client: Redisearch,
  repos: Record<string, RedisRepository<any>>,
  option?: { debug?: boolean; notifyExpiryBySec?: number }
) => QueryDatabaseV2 = (
  client,
  repos,
  { debug, notifyExpiryBySec } = { debug: false, notifyExpiryBySec: 86400 }
) => {
  const logger = getLogger({ name: '[query-handler] createQueryDatabase.js', target: 'console' });
  const commitRepo = createRedisRepository<Commit, CommitInRedis, OutputCommit>({
    client,
    kind: 'commit',
    fields: commitSearchDefinition,
    preSelector,
    postSelector,
    entityName: 'commit',
  });

  const notificationCenter = createNotificationCenter(client);

  // add built-in commit repo
  const allRepos = Object.assign({}, repos, { commit: commitRepo });

  const getHistory = (commits: (Commit | OutputCommit)[]): any[] => {
    const history = [];
    commits.forEach(({ events }) => events.forEach((event) => history.push(event)));
    return history;
  };

  const deleteItems = async <TItem>(repo: RedisRepository<TItem>, pattern: string) => {
    const [errors, count] = await repo.deleteItemsByPattern(pattern);
    const isError = errors?.reduce((pre, cur) => pre || !!cur, false);

    return isError
      ? {
          status: 'ERROR' as any,
          message: `${count} record(s) deleted`,
          errors,
        }
      : {
          status: 'OK' as any,
          message: `${count} record(s) deleted`,
          data: count,
        };
  };

  const queryCommit = async (pattern, args) => {
    const [errors, data] = await commitRepo.queryCommitsByPattern(
      commitRepo.getPattern(pattern, args)
    );
    const isError = errors?.reduce((pre, cur) => pre || !!cur, false);

    debug && console.debug(util.format('returns data, %j', data));
    debug && console.debug(util.format('returns error, %j', errors));

    return isError
      ? { status: 'ERROR' as any, message: QUERY_ERR, errors }
      : { status: 'OK' as any, message: `${data.length} record(s) returned`, data };
  };

  const doSearch: <T>(option: {
    repo: RedisRepository<T>;
    kind: 'commit' | 'entity';
    query: string;
    param: FTSearchParameters;
    countTotalOnly: boolean;
  }) => Promise<HandlerResponse> = async ({ repo, kind, query, param, countTotalOnly }) => {
    const { search, getIndexName } = repo;
    const index = getIndexName();
    const [errors, count, data] = await search({
      countTotalOnly,
      kind,
      index,
      query,
      param,
      restoreFn: kind === 'entity' && repo.getPostSelector(),
    });
    const isError = errors?.reduce((pre, cur) => pre || !!cur, false);

    debug && console.debug(util.format('returns data, %j', data));
    debug && console.debug(util.format('returns error, %j', errors));

    return isError
      ? { status: 'ERROR', message: 'search error', errors }
      : {
          status: 'OK',
          message: `${count} record(s) returned`,
          data: countTotalOnly ? count : data,
        };
  };

  return {
    clearNotification: async (option) => notificationCenter.clearNotification(option),
    clearNotifications: async (option) => notificationCenter.clearNotifications(option),
    deleteCommitByEntityId: async ({ entityName, id }) => {
      if (!entityName || !id) throw new Error(INVALID_ARG);

      return deleteItems(
        commitRepo,
        commitRepo.getPattern('COMMITS_BY_ENTITYNAME_ENTITYID', [entityName, id])
      );
    },
    deleteCommitByEntityName: async ({ entityName }) => {
      if (!entityName) throw new Error(INVALID_ARG);

      return deleteItems(commitRepo, commitRepo.getPattern('COMMITS_BY_ENTITYNAME', [entityName]));
    },
    deleteEntityByEntityName: async <TEntity>({ entityName }) => {
      if (!entityName) throw new Error(INVALID_ARG);
      const entityRepo = allRepos[entityName];

      if (!entityRepo) throw new Error(REPO_NOT_FOUND);

      return deleteItems<TEntity>(
        entityRepo,
        entityRepo.getPattern('ENTITIES_BY_ENTITYNAME', [entityName])
      );
    },
    fullTextSearchCommit: async ({ query, param, countTotalOnly }) => {
      if (!query) throw new Error(INVALID_ARG);

      return doSearch<OutputCommit>({
        repo: commitRepo,
        countTotalOnly,
        kind: 'commit',
        query,
        param,
      });
    },
    fullTextSearchEntity: async <TEntity>({ entityName, query, param, countTotalOnly }) => {
      if (!query || !entityName) throw new Error(INVALID_ARG);

      const repo = allRepos[entityName];
      if (!repo) throw new Error(REPO_NOT_FOUND);

      return doSearch<TEntity>({ repo, countTotalOnly, kind: 'entity', query, param });
    },
    getNotification: async ({ creator, entityName, id, commitId }) =>
      notificationCenter.getNotification({ creator, entityName, id, commitId }),
    getNotificationsByFields: async ({ creator, entityName, id }) =>
      notificationCenter.getNotificationsByFields({ creator, entityName, id }),
    getRedisCommitRepo: () => commitRepo,
    queryCommitByEntityId: async ({ entityName, id }) => {
      if (!entityName || !id) throw new Error(INVALID_ARG);

      return queryCommit('COMMITS_BY_ENTITYNAME_ENTITYID', [entityName, id]);
    },
    queryCommitByEntityName: async ({ entityName }) => {
      if (!entityName) throw new Error(INVALID_ARG);

      return queryCommit('COMMITS_BY_ENTITYNAME', [entityName]);
    },
    mergeCommit: async ({ commit }) => {
      if (!isCommit(commit)) throw new Error(INVALID_ARG);

      try {
        const key = allRepos['commit'].getKey(commit);
        const status = await allRepos['commit'].hmset(commit);
        return {
          status,
          message: `${key} merged successfully`,
          data: [key],
        };
      } catch (e) {
        logger.error(util.format('mergeCommit - %s, %j', REDIS_ERR, e));
        throw e;
      }
    },
    mergeCommitBatch: async ({ entityName, commits }) => {
      if (!entityName || !commits) throw new Error(INVALID_ARG);
      if (isEqual(commits, {}))
        return {
          status: 'OK',
          message: NO_RECORDS,
          data: [],
        };

      const data = [];
      const error = [];
      try {
        for await (const commit of Object.values(commits)) {
          const status = await allRepos['commit'].hmset(commit);
          const key = allRepos['commit'].getKey(commit);
          if (status === 'OK') data.push(key);
          else error.push(key);
        }
      } catch (e) {
        logger.error(util.format('mergeCommitBatch - %s, %j', REDIS_ERR, e));
        throw e;
      }
      debug && console.debug(util.format('data returns: %j', data));
      debug && console.debug(util.format('error returns: %j', error));

      return {
        status: error.length === 0 ? 'OK' : 'ERROR',
        message: `${data.length} record(s) merged successfully`,
        data,
        error,
        errors: error,
      };
    },
    mergeEntity: async ({ commit, reducer }) => {
      const { entityName, entityId, commitId } = commit;
      const entityRepo = allRepos[entityName];
      const entityKeyInRedis = allRepos[entityName].getKey(commit);
      const commitKeyInRedis = allRepos['commit'].getKey(commit);

      if (!entityRepo) throw new Error(REPO_NOT_FOUND);
      if (!isCommit(commit) || !reducer) throw new Error(INVALID_ARG);

      // step 1: retrieve existing commit
      const pattern = commitRepo.getPattern('COMMITS_BY_ENTITYNAME_ENTITYID', [
        entityName,
        entityId,
      ]);

      const [errors, restoredCommits] = await commitRepo.queryCommitsByPattern(pattern);
      const isError = errors?.reduce((pre, cur) => pre || !!cur, false);

      debug && console.debug('restored commits, %j', restoredCommits);

      if (isError)
        return {
          status: 'ERROR',
          message: 'fail to retrieve existing commit',
          errors,
        };

      // step 2: merge existing record with newly retrieved commit
      const history: (Commit | OutputCommit)[] = [...restoredCommits, commit];

      // step 3: compute the timeline of event history
      const state = reducer(getHistory(history));
      /* e.g. newly computed state =
      {
        id: 'qh_proj_test_001',
        desc: 'query handler #2 proj',
        tag: 'projection',
        value: 2,
        _ts: 1590739000,
        _created: 1590738792,
        _creator: 'org1-admin'
      }
      */

      // step 4 add Tracking Info
      // TODO: need Paul's help about how to represent tracking information in Redis
      // const newComputedState = Object.assign({}, state, trackingReducer(history));

      debug && console.debug(util.format('entity being merged, %j', state));

      // step 5: compute events history, returning comma separator
      if (!state?.id) {
        return {
          status: 'ERROR',
          message: REDUCE_ERR,
          errors: [new Error(`fail to reduce, ${entityName}:${entityId}:${commitId}`)],
        };
      }

      const data = [];
      try {
        let status;
        // step 6: add entity
        status = await allRepos[entityName].hmset(state, history);
        data.push({ key: entityKeyInRedis, status });

        // step 7: add commit
        status = await allRepos['commit'].hmset(commit);
        data.push({ key: commitKeyInRedis, status });

        // step 8: add notification flag
        await notificationCenter.notify({
          creator: state._creator,
          entityName,
          id: entityId,
          commitId,
        });
      } catch (e) {
        if (!e.message.startsWith('[lifecycle]'))
          logger.error(util.format('mergeEntity - %s, %j', REDIS_ERR, e));
        throw e;
      }

      debug && console.debug(util.format('data returns: %j', data));

      return { status: 'OK', message: `${entityKeyInRedis} merged successfully`, data };
    },
    mergeEntityBatch: async ({ entityName, commits, reducer }) => {
      if (!entityName || !commits || !reducer) throw new Error(INVALID_ARG);

      const entityRepo = allRepos[entityName];
      if (!entityRepo) throw new Error(REPO_NOT_FOUND);

      if (isEqual(commits, {}))
        return {
          status: 'OK',
          message: NO_RECORDS,
          data: [],
        };

      // safety filter: ensure only relevant entityName is processed
      const filteredCommits = filter(commits, { entityName });
      const groupByEntityId: Record<string, Commit[]> = groupBy(filteredCommits, ({ id }) => id);
      const errors = [];
      const entities = Object.entries(groupByEntityId)
        .map(([entityId, commits]) => {
          const state = reducer(getHistory(commits));
          const keyOfEntityInRedis = allRepos[entityName].getKey(commits[0]);
          // if reducer fails
          !state && errors.push(entityId);
          return { state, commits, key: keyOfEntityInRedis };
        })
        .filter(({ state }) => !!state); // ensure no null; if error happens when reducing

      debug && console.debug(util.format('errors found, %j', errors));

      // add entity. Notice that the original orginal commit is not saved.
      const data = [];
      for await (const { state, commits, key } of entities) {
        try {
          const status = await allRepos[entityName].hmset(state, commits);
          data.push({ key, status });
        } catch (e) {
          logger.error(util.format('mergeEntityBatch - %s, %j', REDIS_ERR, e));
          throw e;
        }
      }

      debug && console.debug(util.format('data returns: %j', data));

      return {
        status: errors.length === 0 ? 'OK' : 'ERROR',
        message: `${data.length} record(s) merged`,
        data,
        errors: errors.length === 0 ? null : errors,
      };
    },
  };
};
