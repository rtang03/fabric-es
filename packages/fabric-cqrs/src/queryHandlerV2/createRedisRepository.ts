import flatten from 'lodash/flatten';
import { FTCreateParameters, FTSchemaField, Redisearch } from 'redis-modules-sdk';
import type { OutputSelector } from 'reselect';
import type { Commit } from '../types';
import type { FieldOption, RedisearchMapField, RedisRepository } from './types';

/**
 * @about create abstract layer for redis repository
 */
export const createRedisRepository: <TItem, TItemInRedis, TResult>(option: {
  client: Redisearch;
  kind?: 'entity' | 'commit';
  fields: RedisearchMapField<TItem>;
  entityName?: string;
  param?: FTCreateParameters;
  restore?: OutputSelector<TItemInRedis, TResult, any>;
}) => RedisRepository<TResult> = <TItem, TItemInRedis, TResult>({
  client,
  kind = 'entity' as any,
  fields,
  entityName,
  param,
  restore,
}) => {
  // every entity is indexed with Prefix "e:entityName:". commit is "c:"
  const indexName = { entity: `eidx`, commit: 'cidx' }[kind];

  const prefix = { entity: `e:`, commit: 'c:' }[kind];

  // compute key
  const getKey = {
    entity: ({ entityName, entityId }) => `${prefix}${entityName}:${entityId}`,
    commit: ({ entityName, entityId, commitId }: Commit) =>
      `${prefix}${entityName}:${entityId}:${commitId}`,
  }[kind];

  // convert to Redis Hash fields format, before hset / hmset
  const transformBeforeHset: <E>(fields: RedisearchMapField<E>, item: E) => (string | number)[] = <
    E
  >(
    input,
    item
  ) =>
    flatten<string | number>(
      Object.entries<FieldOption<E>>(input).map(([key, { altName, transform }]) => [
        // if alternate name exist, will replace orgingal key
        altName ?? key,
        // if transformation rule not exist, use original value
        transform?.(item) || item[key],
      ])
    );

  // add default FTCreateParameters
  const getParam: (param: FTCreateParameters) => FTCreateParameters = (param) =>
    Object.assign({}, { prefix: [{ count: 1, name: prefix }] }, param);

  // compute schema
  const getSchema: <F>(fields: F) => FTSchemaField[] = <F extends FieldOption>(input) =>
    Object.entries<F>(input)
      .map(
        ([key, { altName, index }]) =>
          // if index is required
          index && {
            ...index,
            // append the indexKey name
            name: altName ?? key,
          }
      )
      .filter((item) => !!item);

  return {
    createIndex: () => client.create(indexName, getSchema<TItem>(fields), getParam(param)),
    dropIndex: (deleteHash = true) => client.dropindex(indexName, deleteHash),
    hmset: (item) => client.redis.hmset(getKey(item), transformBeforeHset<TItem>(fields, item)),
    hgetall: (key) =>
      client.redis.hgetall(key).then((result) => (restore?.(result) || result) as TResult),
    getKey: (item) => getKey(item),
    getIndexName: () => indexName,
    convert: (item) => transformBeforeHset<TItem>(fields, item),
  };
};
