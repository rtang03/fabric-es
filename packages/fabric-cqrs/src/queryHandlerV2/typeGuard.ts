import type { OutputCommit } from './types';

export const isOutputCommit = (value: any): value is OutputCommit =>
  value?.id !== undefined &&
  value?.entityName !== undefined &&
  value?.commitId !== undefined &&
  value?.mspId !== undefined &&
  value?.event !== undefined &&
  value?.entityId !== undefined &&
  typeof value?.version === 'number' &&
  typeof value?.ts === 'number';
