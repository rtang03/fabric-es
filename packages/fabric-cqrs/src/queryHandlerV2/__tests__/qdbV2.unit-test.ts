require('dotenv').config({ path: './.env.dev' });
import omit from 'lodash/omit';
import pick from 'lodash/pick';
import { Redisearch } from 'redis-modules-sdk';
import {
  Counter,
  reducer,
  counterSearchDefinition as fields,
  CounterInRedis,
  postSelector,
  preSelector,
  OutputCounter,
} from '../../unit-test-reducer';
import { REDUCE_ERR } from '../constants';
import { createQueryDatabaseV2 } from '../createQueryDatabaseV2';
import { createRedisRepository } from '../createRedisRepository';
import { isOutputCommit } from '../typeGuard';
import type { QueryDatabaseV2, RedisRepository, OutputCommit } from '../types';
import { commit, commits, faultReducer, newCommit } from './__utils__';

// const key = `${commit.entityName}::${commit.entityId}::${commit.commitId}`;
// const key2 = `test_proj::qh_proj_test_002`;
// const key3 = `test_proj::qh_proj_test_003`;

/**
 * running it: .dn-run.0-db-red.sh
 */
let queryDatabase: QueryDatabaseV2;
let client: Redisearch;
let commitRepo: RedisRepository<OutputCommit>;
let counter: RedisRepository<OutputCounter>;

const ENTITYNAME = 'test_proj';
const ENTITYID = 'qh_proj_test_001';
const noResult = { status: 'OK', message: '0 record(s) returned', result: [] };

beforeAll(async () => {
  client = new Redisearch({ host: 'localhost', port: 6379 });

  await client.connect();

  counter = createRedisRepository<Counter, CounterInRedis, OutputCounter>({
    client,
    fields,
    entityName: 'test_proj',
    postSelector,
    preSelector,
  });

  queryDatabase = createQueryDatabaseV2(client, { [ENTITYNAME]: counter });
  commitRepo = queryDatabase.getRedisCommitRepo();

  // prepare eidx
  const eidx = counter.getIndexName();
  await counter
    .dropIndex()
    .then((result) => console.log(`${eidx} is dropped: ${result}`))
    .catch((result) => console.log(`${eidx} is not dropped: ${result}`));

  await counter
    .createIndex()
    .then((result) => console.log(`${eidx} is created: ${result}`))
    .catch((result) => {
      console.log(`${eidx} is not created: ${result}`);
      process.exit(1);
    });

  const cidx = commitRepo.getIndexName();

  await commitRepo
    .dropIndex()
    .then((result) => console.log(`${cidx} is dropped: ${result}`))
    .catch((result) => console.log(`${cidx} is not dropped: ${result}`));

  await commitRepo
    .createIndex()
    .then((result) => console.log(`${cidx} is created: ${result}`))
    .catch((result) => {
      console.log(`${cidx} is not created: ${result}`);
      process.exit(1);
    });
});

afterAll(async () => {
  await client.disconnect();
  return new Promise<void>((ok) => setTimeout(() => ok(), 2000));
});

describe('Projecion db test', () => {
  it('should deleteCommitByEntityName', async () =>
    queryDatabase
      .deleteCommitByEntityName({ entityName: ENTITYNAME })
      .then(({ status }) => expect(status).toBe('OK')));

  // first commit for merge test
  it('should merge commit', async () => {
    const key = commitRepo.getKey(commit);
    // test the returned result
    const { result, status } = await queryDatabase.mergeCommit({ commit });
    expect(status).toBe('OK');
    expect(result).toStrictEqual([key]);

    // test the reselected data
    const data = await commitRepo.hgetall(key);
    expect(omit(commit, 'events')).toStrictEqual(
      pick(data, 'id', 'entityName', 'version', 'commitId', 'entityId', 'mspId')
    );
  });

  it('should fail to mergeEntity with IRRELEVANT reducer', async () =>
    queryDatabase
      .mergeEntity({ commit: newCommit, reducer: faultReducer })
      .then(({ status, message, error }) => {
        expect(status).toEqual('ERROR');
        expect(error.message.startsWith('fail to reduce')).toBeTruthy();
        expect(message).toContain(REDUCE_ERR);
      }));

  it('should merge entity', async () =>
    queryDatabase
      .mergeEntity({ commit: newCommit, reducer })
      .then(({ status, message, result }) => {
        expect(status).toBe('OK');
        expect(result).toEqual([
          { key: 'e:test_proj:qh_proj_test_001', status: 'OK' },
          {
            key: 'c:test_proj:qh_proj_test_001:20200528133520841',
            status: 'OK',
          },
        ]);
      }));

  it('should fail to queryCommitByEntityId: non-exist entityName', async () =>
    queryDatabase
      .queryCommitByEntityId({ entityName: 'ABC', id: ENTITYID })
      .then((result) => expect(result).toEqual(noResult)));

  it('should fail to queryCommitByEntityId: non-exist id', async () =>
    queryDatabase
      .queryCommitByEntityId({ entityName: ENTITYNAME, id: 'DEF' })
      .then((result) => expect(result).toEqual(noResult)));

  it('should queryCommitsBy: entityName and entityId', async () =>
    queryDatabase
      .queryCommitByEntityId({ entityName: ENTITYNAME, id: ENTITYID })
      .then(({ status, message, result }) => {
        expect(status).toBe('OK');
        expect(message).toEqual('2 record(s) returned');
        result.forEach((commit) => expect(isOutputCommit(commit)).toBeTruthy());
      }));

  it('should fail to queryCommitsBy: non-exist entityName', async () =>
    queryDatabase
      .queryCommitByEntityName({ entityName: 'ABC' })
      .then((result) => expect(result).toEqual(noResult)));

  it('should queryCommitsBy: entityName', async () =>
    queryDatabase
      .queryCommitByEntityName({ entityName: ENTITYNAME })
      .then(({ status, message, result }) => {
        expect(status).toBe('OK');
        expect(message).toEqual('2 record(s) returned');
        result.forEach((commit) => expect(isOutputCommit(commit)).toBeTruthy());
      }));

  it('should mergeBatch', async () =>
    queryDatabase
      .mergeEntityBatch({ entityName: ENTITYNAME, reducer, commits })
      .then(({ result, status, message }) => {
        expect(status).toBe('OK');
        expect(result).toEqual([
          { key: 'e:test_proj:qh_proj_test_002', status: 'OK' },
          { key: 'e:test_proj:qh_proj_test_003', status: 'OK' },
        ]);
      }));

  it('should fullTextSearchCommit: qh*', async () =>
    queryDatabase.fullTextSearchCommit({ query: 'qh*' }).then(({ status, result }) => {
      expect(status).toBe('OK');
      (result as any[]).forEach((item) => expect(isOutputCommit(item)).toBeTruthy());
      expect((result as any[]).length).toBe(2);
    }));

  it('should fullTextSearchCommit: return count-only', async () =>
    queryDatabase
      .fullTextSearchCommit({ query: 'qh*', countTotalOnly: true })
      .then(({ status, result }) => {
        expect(status).toBe('OK');
        expect(result).toBe(2);
      }));

  it('should fail to fullTextSearchEntity: invalid character', async () =>
    queryDatabase
      .fullTextSearchEntity({ entityName: ENTITYNAME, query: 'xffd;;;;;df' })
      .then(({ status, error }) => {
        expect(status).toEqual('ERROR');
        // "error": [Error: Redisearch: ReplyError: Syntax error at offset 4 near xffd],
      }));

  it('should fail to fullTextSearchEntity: entityName not found', async () =>
    queryDatabase.fullTextSearchEntity({ entityName: ENTITYNAME, query: 'abc' }).then((result) => {
      expect(result).toEqual({ status: 'OK', message: '0 record(s) returned', result: [] });
    }));

  it('should fail to fullTextSearchEntity by "de": invalid input', async () =>
    queryDatabase
      .fullTextSearchEntity({
        query: 'xyz',
        entityName: ENTITYNAME,
        param: { sortBy: { sort: 'ASC', field: 'de' } },
      })
      .then((result) => expect(result).toEqual(noResult)));

  it('should fail to fullTextSearchEntity by "abc": no such indexed field', async () =>
    queryDatabase
      .fullTextSearchEntity({
        query: 'xyz',
        entityName: ENTITYNAME,
        param: { sortBy: { sort: 'ASC', field: 'abc' } },
      })
      .then(({ status, error }) => {
        expect(status).toBe('ERROR');
        expect(error.message).toContain('not loaded nor in schema');
      }));

  it('should fullTextSearchEntity by "de"', async () =>
    queryDatabase
      .fullTextSearchEntity({
        query: 'handler',
        entityName: ENTITYNAME,
        param: { sortBy: { sort: 'ASC', field: 'de' } },
      })
      .then((result) => expect(result).toMatchSnapshot()));

  it('should fullTextSearchEntity by "de": countTotalOnly', async () =>
    queryDatabase
      .fullTextSearchEntity({
        countTotalOnly: true,
        query: 'handler',
        entityName: ENTITYNAME,
        param: { sortBy: { sort: 'ASC', field: 'de' } },
      })
      .then((result) =>
        expect(result).toEqual({ status: 'OK', message: '3 record(s) returned', result: 3 })
      ));

  // it('should fail to deleteCommitByEntityId with invalid entityName', async () =>
  //   queryDatabase
  //     .deleteCommitByEntityId({ entityName: 'ABC', id: ENTITYID })
  //     .then((result) =>
  //       expect(result).toEqual({ status: 'OK', message: '0 record(s) deleted', result: 0 })
  //     ));
  //
  // it('should deleteCommitByEntityId', async () =>
  //   queryDatabase
  //     .deleteCommitByEntityId({ entityName: ENTITYNAME, id: ENTITYID })
  //     .then((result) =>
  //       expect(result).toEqual({ status: 'OK', message: '2 record(s) deleted', result: 2 })
  //     ));
});
