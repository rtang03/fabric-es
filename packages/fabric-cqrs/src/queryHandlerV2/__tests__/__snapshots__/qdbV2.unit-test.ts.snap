// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projecion db test should fullTextSearchEntity by "de" 1`] = `
Object {
  "message": "3 record(s) returned",
  "result": Array [
    Object {
      "created": "1590738792",
      "creator": "org1-admin",
      "de": "query handler #2 proj",
      "event": "Increment,Increment",
      "history": "c:test_proj:qh_proj_test_00120200528133519841,c:test_proj:qh_proj_test_00120200528133520841",
      "id": "qh_proj_test_001",
      "tag": "projection",
      "tl": "1590738792,1590739000",
      "ts": "1590739000",
      "val": "2",
    },
    Object {
      "created": "1590740000",
      "creator": "org1-admin",
      "de": "query handler #5 proj",
      "event": "Increment,Increment,Increment",
      "history": "c:test_proj:qh_proj_test_00220200528133530001,c:test_proj:qh_proj_test_00220200528133530002,c:test_proj:qh_proj_test_00220200528133530003",
      "id": "qh_proj_test_002",
      "tag": "projection",
      "tl": "1590740000,1590740001,1590740002",
      "ts": "1590740002",
      "val": "3",
    },
    Object {
      "created": "1590740003",
      "creator": "org1-admin",
      "de": "query handler #7 proj",
      "event": "Increment,Increment",
      "history": "c:test_proj:qh_proj_test_00320200528133530004,c:test_proj:qh_proj_test_00320200528133530005",
      "id": "qh_proj_test_003",
      "tag": "projection",
      "tl": "1590740003,1590740004",
      "ts": "1590740004",
      "val": "2",
    },
  ],
  "status": "OK",
}
`;
